File: frontend/formfilling/pr/fill.js
/* This is the main entry point for all page forms filling. */
// const { Playwright } = require('playwright'); \
//TODO:  the model page /pages may be changed, so all the pages need to be updated

const WebPages = require('../pages');
const { Login, ViewApplication, ApplicationPicker, FormPicker } = require('./poms/start');
const { Imm0008Intro, ApplicationDetail, PersonalDetail, ContactInformation, Passport, NationalID, EducationOccupation, LanguageDetail } = require('./poms/imm0008');
// const { timing } = require('../../utils/utils/timing');
const Filler = require('../filler');
const path = require('path');

async function buildPages(page, args) {
    /** Build pages according to page, page names, and data */

    const loginPage = new Login(page, 'login', "Log in", null);
    const viewPage = new ViewApplication(page, 'view_application', "Start new application or view my PR application", null);
    const picker = new ApplicationPicker(page, 'picker', "Pick one's application by email", null);
    const formPicker = new FormPicker(page, 'form_picker', "Pick a form to be filled", null);
    const imm0008Intro = new Imm0008Intro(page, 'imm0008_intro', "Imm0008 form filling introduction", null);
    const applicationDetail = new ApplicationDetail(page, 'application_detail', "Application details", null);
    const personalDetail = new PersonalDetail(page, 'personal_detail', "Personal details", null);
    const contactInformation = new ContactInformation(page, 'contact_information', "Contact information", null);
    const passport = new Passport(page, 'passport', "Passport details", null);
    const nationalId = new NationalID(page, 'national_id', "National Id details", null);
    const educationOccupation = new EducationOccupation(page, 'education_occupation', "Educations and occupations", null);
    const languageDetail = new LanguageDetail(page, 'language_detail', "Language details", null);

    const webPages = new WebPages(
        [
            loginPage,
            viewPage,
            picker,
            formPicker,
            imm0008Intro,
            applicationDetail,
            personalDetail,
            contactInformation,
            passport,
            nationalId,
            educationOccupation,
            languageDetail
        ]
    );

    return webPages.pages;
}

async function main() {
    const sourceExcel = path.join(__dirname, '/output/test.xlsx');
    const skipPagesRange = null;

    const args = {
        sourceExcel,
        pdf: false,
        png: false,
        folder: path.join(__dirname, '/screenshots'),
        skipPagesRange
    };
    filler = new Filler(headless = false, slow_mo = 0);
    await filler.fill(buildPages, args);
}

// timing(main);
main()

File: client/webform/pr/fill.py
""" This is the main entry point for all page forms filling. """

from fileinput import filename
from playwright.sync_api import Playwright, sync_playwright
from client.webform.pages import WebPages
import json, os
from typing import Union
from client.system.config import console
from base.utils.client.utils import timing
from base.source.excel import Excel
from client.webform.filler import Filler
from client.webform.pr.poms.start import Login,ViewApplication,ApplicationPicker,FormPicker
from client.webform.pr.poms.imm0008 import Imm0008Intro,ApplicationDetail,PersonalDetail,ContactInformation,Passport,NationalID,EducationOccupation

def build_pages(page,*,source_excel=None,**args):
    """Build pages according to page, page names, and data"""

    login_page = Login(name="login", page=page, data=None)
    view_page = ViewApplication(name="view_application", page=page, data=None)
    picker=ApplicationPicker(name="picker",page=page,data=None)
    form_picker=FormPicker(name="form_picker",page=page,data=None)
    imm0008_intro=Imm0008Intro(name="imm0008_intro",page=page,data=None)
    application_detail=ApplicationDetail(name="application_detail",page=page,data=None)
    personal_detail=PersonalDetail(name="personal_detail",page=page,data=None)
    contact_information=ContactInformation(name="contact_information",page=page,data=None)
    passport=Passport(name="passport",page=page,data=None)
    nation_id=NationalID(name="national_id",page=page,data=None)
    education_occupation=EducationOccupation(name="education_occupation",page=page,data=None)
    
    web_pages = WebPages(pages=[login_page, view_page,picker,form_picker,imm0008_intro,application_detail,personal_detail,contact_information,passport,nation_id,education_occupation])
    return web_pages.pages


@timing
def main():
    arguments={
        "source_excel":"/Users/jacky/desktop/output/test.xlsx",
        "pdf":False, 
        "png":False, 
        "folder":"/Users/jacky/desktop/screenshots",
        "skip_pages_range":[4,8]
    }
    Filler(headless=False).fill(build_pages,arguments)


if __name__ == "__main__":
    main()